_signal_processing/signal_utils.py

```diff
@@ -0,0 +1,83 @@
+import math
+
+import numpy as np
+from scipy.signal import firwin, lfilter, butter
+
+DEFAULT_NUM_OF_FREQ = 1024
+DEFAULT_FILTER_ORDER = 5
+DEFAULT_FILTER_WINDOW = 'hamming'
+DEFAULT_NUM_TAPS = 256
+
+
+def audible(signal):
+    """
+    Scale so it's audible
+    """
+    signal = np.int16(signal / np.max(np.abs(signal)) * 32767)
+    return signal
+
+
+def butter_lowpass(highcut, fs, order=DEFAULT_FILTER_ORDER):
+    return butter(order, highcut, fs=fs, btype='low')
+
+
+def butter_lowpass_filter(signal, highcut, fs, order=DEFAULT_FILTER_ORDER):
+    b, a = butter_lowpass(highcut, fs, order=order)
+    rx = lfilter(b, a, signal)
+    return rx
+
+
+def firwin_lowpass(highcut, fs, ntaps=DEFAULT_NUM_TAPS, window=DEFAULT_FILTER_WINDOW):
+    taps = firwin(ntaps, highcut,
+                  fs=fs, window=window, scale=False)
+    return taps
+
+
+def firwin_lowpass_filter(data, highcut, fs, ntaps=DEFAULT_NUM_TAPS, window=DEFAULT_FILTER_WINDOW):
+    taps = firwin_lowpass(highcut, fs, ntaps, window)
+    y = lfilter(taps, 1, data)
+    return y
+
+
+def audio_agc(samples, sample_freq, t_max):
+    """
+    Audio Adaptive Gain Control.
+    Primitive version of AGC with slicing samples by time and rescaling each time-slice
+    """
+    t_slice = int(t_max * sample_freq)
+    demod_channel = samples
+
+    new_demod_channel = None
+    demod_channel_chunks = np.array_split(demod_channel, math.ceil(demod_channel.size / t_slice))
+    for chunk in demod_channel_chunks:
+        aud_chunk = audible(chunk)
+        if new_demod_channel is not None:
+            new_demod_channel = np.concatenate((new_demod_channel, aud_chunk))
+        else:
+            new_demod_channel = aud_chunk
+
+    return audible(new_demod_channel)
+
+
+def fm_demod(samples, sample_freq, freq_width):
+    demod_channel = samples
+
+    demod_channel = firwin_lowpass_filter(demod_channel, int(freq_width/2), sample_freq)
+
+    baseband = demod_channel
+    demod_channel = np.angle(baseband[1::1] * np.conjugate(baseband[0:-1:1]))
+    demod_channel = (demod_channel * sample_freq) / (2 * np.pi)
+
```

Comment:
You might need to add deemphasis filter here.

Reply:
Actually, I played with it and have not noticed any differences